core/src/main/scala/org/apache/spark/ContextCleaner.scala

```diff
@@ -58,7 +58,12 @@ private class CleanupTaskWeakReference(
  */
 private[spark] class ContextCleaner(sc: SparkContext) extends Logging {
 
-  private val referenceBuffer = new ConcurrentLinkedQueue[CleanupTaskWeakReference]()
+  /**
+   * A buffer to ensure that `CleanupTaskWeakReference`s are not garbage collected as long as they
+   * have not been handled by the reference queue.
+   */
+  private val referenceBuffer =
+    new ConcurrentHashMap[CleanupTaskWeakReference, CleanupTaskWeakReference]()
```

Comment:
> on its face it sounds less efficient to use a Map as a Set.

`HashSet` is basically just a wrapper of `HashMap`. It's almost same as `Collections.newSetFromMap`. And `ConcurrentHashMap` is better than `synchronizedSet` is because it uses per-bin locks and doesn't lock on the whole map.